File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const n = maps.length;
    const m = maps[0].length;
    const dx = [-1, 0, 1, 0];
    const dy = [0, 1, 0, -1];
    
    const queue = [[0, 0]];
    const visited = Array.from(Array(n), () => Array(m).fill(false));
    visited[0][0] = true;
    
    const distance = Array.from(Array(n), () => Array(m).fill(0));
    distance[0][0] = 1;
    
    while (queue.length>0) {
        const [curX, curY] = queue.shift();
        for (let i=0; i<4; i++){
            const x = curX + dx[i];
            const y = curY + dy[i];
            
            if (x>=0 && x<n && y>=0 && y<m && maps[x][y]===1 && !visited[x][y]) {
                    queue.push([x,y]);
                    visited[x][y] = true;
                    distance[x][y] = distance[curX][curY] + 1; 
            }
        }
    }
    
    return distance[n-1][m-1] > 0 ? distance[n-1][m-1]:-1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# 문제 분석
'''
진영을 파괴하기 위해 진영한테 제일 빨리 도착하는 방법
진영한테 도착하지 못하는 경우는 -1
'''

# 의사 결정
'''
최단 거리를 구하는 문제이므로 BFS 활용
'''
from collections import deque

def bfs(start, maps, visited, directions, n, m):
    queue = deque([(start[0], start[1], 1)])
    visited[start[0]][start[1]] = 1
    
    while queue:
        x,y,count = queue.popleft()
        if x == n and y == m:
            return count
        
        for dx, dy in directions:
            nx, ny = x+dx, y+dy
            if 1 <= nx <= n and 1 <= ny <= m:
                if not visited[nx][ny] and maps[nx-1][ny-1] == 1:
                    visited[nx][ny] = 1
                    queue.append((nx, ny, count+1))
    return -1

    

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    visited = [[0]*(m+1) for _ in range(n+1)]
    directions = [(1,0),(0,1),(-1,0),(0,-1)]
    return bfs((1,1), maps, visited, directions, n, m) 
    